### PR TITLE
fix(Core/Movement): Improvements to taxi flight movement generator:

### LINKED ISSUE #7842: Logging into a character who is mid-flight can cause you to not join general chat for that zone
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [ ] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour

I took a flight path from Stormwind to Lakeshire, and logged off of the character shortly into the flight. When I logged back in, on the bird, my character did not join general chat for either Elwynn Forest or for Stormwind (nor trade chat). See screenshot.

This does seem like a bug, but it also seems harmless. If you're flying, chances are you will be arriving at a new zone soon enough anyway and join that general channel.

### Expected Blizzlike Behaviour

Even on a flightpath, when you log in to a character, you should join the general chat of the zone your character is in, even while they are flying over it.

### Source

_No response_

### Steps to reproduce the problem

Hop on a flight path.
Log off, perhaps it needs to be soon after crossing the border into another zone.
Log back on, see if you are entered into general chat for the zone.

### Extra Notes

You can see the yellow text messages showing that I just logged on, and the white text of the travelling merchant NPC who patrols on the ground along the path that the flight path takes from Stormwind to Goldshire, and you can see that my character joins world chat as /1, and no other chat.
Screenshot (60)

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/595bb6adccbabc714469f3935541978283b8bdfb

### Operating system

Windows 10

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Customizations

None

### Server

ChromieCraft

https://github.com/chromiecraft/chromiecraft/issues/1711

### LINKED ISSUE #12341: Flight Paths are not blizzlike.
https://github.com/chromiecraft/chromiecraft/issues/3670

Asking for feedback because I thought flight path stuff came from Blizzard Files.

### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [ ] 60

### Current Behaviour

Chained Flight Paths make a stop at each flight master that makes part of the chain.

### Expected Blizzlike Behaviour

Chained Flight Paths should skip a few waypoints and not stop at any flight master besides the first and the last of the chain.

### Source

Several youtube videos from the WotLK era. See Extra Notes.

### Steps to reproduce the problem

Take a long flightpath over multiple nodes

### Extra Notes

First video is retail, second is AzerothCore

<details>
 <summary>Flight 1: Lakeshire -> Stormwind -> Ironforge</summary>

https://www.youtube.com/watch?v=8TBaZ37q5EY

Lakeshire -> **Stormwind** -> Ironforge

https://user-images.githubusercontent.com/100873791/178131486-f20fa2f0-adf9-4da7-bebf-1fb585883a72.mp4

https://user-images.githubusercontent.com/100873791/178131479-ff90de69-21ae-41b4-85c5-500da3b5dd9f.mp4

</details>

<details>
 <summary>Flight 2: Westguard Keep -> Wintergarde Keep -> Stars' Rest -> Valiance Keep</summary>

https://www.youtube.com/watch?v=s01aKPWByMg

Westguard Keep -> Wintergarde Keep -> **Stars' Rest** -> Valiance Keep

https://user-images.githubusercontent.com/100873791/178131645-ef7f9e9c-7f38-4ea0-bc7d-645aea9cb228.mp4

https://user-images.githubusercontent.com/100873791/178131647-cfef2820-63c6-4f27-8e2c-993ea2f7ebb6.mp4

</details>

<details>
 <summary>Flight 3: Ironforge -> Stormwind City -> Booty Bay</summary>

https://www.youtube.com/watch?v=LsAJor7oayY

Ironforge -> **Stormwind City** -> Booty Bay

https://user-images.githubusercontent.com/100873791/178131733-f898573d-5135-40eb-80c5-9dbccc7ec496.mp4

https://user-images.githubusercontent.com/100873791/178131734-1a0109b2-11d3-4231-87f8-2d049b6179ac.mp4

</details>

<details>
 <summary>Flight 4: Rut'theran Village -> Auberdine -> Theramore Isle -> Gadgetzan</summary>

https://www.youtube.com/watch?v=-7bBxW4a82w

Rut'theran Village -> **Auberdine** -> Theramore Isle -> Gadgetzan

https://user-images.githubusercontent.com/100873791/178131783-6e013a3a-51c0-424b-af77-79d7d8932774.mp4

https://user-images.githubusercontent.com/100873791/178131793-1258ad1a-fb5d-4ce9-870c-de15385e598a.mp4

Rut'theran Village -> Auberdine -> **Theramore Isle** -> Gadgetzan

https://user-images.githubusercontent.com/100873791/178131790-5a9c4afe-e03b-4e71-a68c-cce3f1e51a2e.mp4

https://user-images.githubusercontent.com/100873791/178131786-cfec8aa6-69d9-41f4-9d50-437064bfb1f4.mp4

</details>

<details>
 <summary>Flight 5: Moonglade -> Bloodvenom Post -> Zoram'gar Outpost -> Sun Rock Retreat -> Shadowprey Village</summary>

https://www.youtube.com/watch?v=t854bUR32Hc

Moonglade -> **Bloodvenom Post** -> Zoram'gar Outpost -> Sun Rock Retreat -> Shadowprey Village

https://user-images.githubusercontent.com/100873791/178131913-c8c07102-8831-47b3-ae9a-0bbaac1e410d.mp4

https://user-images.githubusercontent.com/100873791/178131920-c5353044-93a5-47b1-bd81-2cad05277ece.mp4

Moonglade -> Bloodvenom Post -> **Zoram'gar Outpost** -> Sun Rock Retreat -> Shadowprey Village

https://user-images.githubusercontent.com/100873791/178131924-2d57a7ad-d64c-4689-aa35-dc212e986cd1.mp4

https://user-images.githubusercontent.com/100873791/178131927-458c6d6f-cf8e-487b-bf09-354166768594.mp4

Moonglade -> Bloodvenom Post -> Zoram'gar Outpost -> **Sun Rock Retreat** -> Shadowprey Village

https://user-images.githubusercontent.com/100873791/178131929-17a246ec-7b66-44d5-bfb1-b5b7e984a155.mp4

https://user-images.githubusercontent.com/100873791/178131930-3c397c03-c8a2-450e-a451-6eaf8c02cefc.mp4

</details>

<details>
 <summary>Flight 6: Dalaran -> Fordragon Hold -> Stars' Rest -> (...)</summary>

https://www.youtube.com/watch?v=1ZjgJbcGCRM

Flight 6: Dalaran -> **Fordragon Hold** -> Stars' Rest -> (...)

https://user-images.githubusercontent.com/100873791/178131995-8b903419-8705-4d9b-8daa-3d24459e5ba7.mp4

https://user-images.githubusercontent.com/100873791/178131996-5be22998-6b8b-47da-baa4-106f66e2392a.mp4

</details>

<details>
 <summary>Flight 7: Dalaran -> Kor'Kron Vanguard -> Venomspite -> New Agamand</summary>

https://www.youtube.com/watch?v=F52DW3dYdeI

Dalaran -> **Kor'Kron Vanguard** -> Venomspite -> New Agamand

https://user-images.githubusercontent.com/100873791/178132048-b9c76806-ecd7-46a5-bec1-dd5a9aa4fe0c.mp4

https://user-images.githubusercontent.com/100873791/178132050-f09f5181-9874-4a26-8a9d-9959eadbf62e.mp4

Dalaran -> Kor'Kron Vanguard -> **Venomspite** -> New Agamand

https://user-images.githubusercontent.com/100873791/178132054-12debc18-41a1-45b6-a5e8-288b802ab3cf.mp4

https://user-images.githubusercontent.com/100873791/178132059-45d0f792-c926-47bd-a84b-00a773270ed4.mp4

</details>

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/6983903a89c556c0ca1239101cffc38098d6b081

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-dmf-switch](https://github.com/azerothcore/mod-dmf-switch)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna](https://github.com/azerothcore/mod-eluna)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-transmog](https://github.com/azerothcore/mod-transmog)
- [mod-weekend-xp](https://github.com/azerothcore/mod-weekend-xp)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
Changed multi-segment taxi paths to fly nearby flight masters along the way, not directly through them.
Taxi cost on multi-segment paths is now charged per segment when it is started.
Properly send taxi node status on login, as well as if the taxi master is out of range.
Apply reputation discount to all points in multi-segment paths.
Properly clean up list of taxi destinations upon arrival at final node.
Teleport players to the destination taxi node location instead of their current ground position.
Don't start a spline with just 1 point in FlightPathMovementGenerator
Source: TrinityCore.
Fixes #7842
Fixes #12341

## Issues Addressed:

- Closes #12341
- Closes #7842
- Closes https://github.com/chromiecraft/chromiecraft/issues/3670
- Closes https://github.com/chromiecraft/chromiecraft/issues/1711

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

More info [here](#12341) and [here](#7842)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
It's an improvement, although the waypoints aren't 100% blizzlike. But I don't think it's related to this PR.

For example:
 Zoram'gar Outpost -> **Sun Rock Retreat** -> Shadowprey Village: flight will clip through the mountain in Sun Rock Retreat

Kor'kron Vanguard -> **Venomspite** -> New Agamand: will do a very minor clip issue in some of the objects in Venomspite

Moonglade -> **Bloodvenom Post** -> Zoram'gar Outpost: path isn't blizzlike, and does a very small clip in one of the trees.

Other than that I haven't seen any other issues so far. And if the waypoint problems above aren't related to this specific PR I'd say it's good to be merged.

**Comment by acidmanifesto** [comment]
> It's an improvement, although the waypoints aren't 100% blizzlike. But I don't think it's related to this PR.
> 
> For example: Zoram'gar Outpost -> **Sun Rock Retreat** -> Shadowprey Village: flight will clip through the mountain in Sun Rock Retreat
> 
> Kor'kron Vanguard -> **Venomspite** -> New Agamand: will do a very minor clip issue in some of the objects in Venomspite
> 
> Moonglade -> **Bloodvenom Post** -> Zoram'gar Outpost: path isn't blizzlike, and does a very small clip in one of the trees.
> 
> Other than that I haven't seen any other issues so far. And if the waypoint problems above aren't related to this specific PR then it's good to be merged.

can you check leaving from stormwind flight manager to anywhere else and tell me if you hit the wall?

**Comment by Gultask** [comment]
Flying from Stormwind to somewhere else hits no walls from what I saw.
Flying **to** Stormwind also seems to have no problem, although the path takes me farther than what's blizzlike.

If you want me to record it feel free to tell me. I'm not recording because I'm not sure how much it'll help.

**Comment by acidmanifesto** [comment]
> Flying **from** Stormwind to somewhere else hits no walls from what I saw.
> Flying **to** Stormwind also seems to have no problem, although the path takes me farther than what's blizzlike.
> 
> If you want me to record it feel free to tell me. I'm not recording because I'm not sure how much it'll help.

Only ask as some tc variant servers, they usually hit the wall in and out of storm wind.plus with our movement generators being different is anotherthing. With the clip issues mentioned are you going thru the object head on or are you slightly hitting the objects at the tip of the "wings"?

**Comment by Gultask** [comment]
This one is the largest clip. The others I've noticed were just wings clipping due to passing very close to objects.

https://user-images.githubusercontent.com/100873791/178152516-7d2173c3-dec2-4d52-8b86-2d3f79d737fc.mp4

**Comment by acidmanifesto** [comment]
> This one is the largest clip. The others I've noticed were just wings clipping due to passing very close to objects.
> 
>  clip.mp4

haha. i will admit, that full clip thru may be the smoothest one i have ever seen in about 5 servers i worked on. 
I am going to approve this one. Most servers the birdy hits the wall and slides around the facing surface for a few seconds before clipping thru and seeing the whole undermeshing of the map. this is definatelly not that at all.

**Review by acidmanifesto — APPROVED** [review]

**Comment by pklloveyou** [comment]
@Gultask Can you try it?? "Login the game again immediately after the flight, and it will still be in flight"

**Comment by Gultask** [comment]
@mpfans thanks, I'll try that after I'm done with 2 other PRs. I'll report back in 30 mins or less.

File: src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp [hunk 1]
```diff
@@ -313,7 +349,12 @@ void FlightPathMovementGenerator::DoFinalize(Player* player)
     // remove flag to prevent send object build movement packets for flight state and crash (movement generator already not at top of stack)
     player->ClearUnitState(UNIT_STATE_IN_FLIGHT);
 
-    // xinef: this should be cleaned by CleanupAfterTaxiFlight(); function!
+    uint32 taxiNodeId = player->m_taxi.GetTaxiDestination();
+    if (!player->IsInWorld())
```
Comment (pklloveyou) [inline src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp:353]: if (!player->IsInWorld()) Is this wrong?

Comment (pklloveyou) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp:353]: After the flight is completed and lands, the error I mentioned above is caused by the error here

**Comment by Gultask** [comment]
I believe this is what pkIloveyou meant?

https://user-images.githubusercontent.com/100873791/178154725-2bb2dc5e-ca84-44df-92a7-1019691cf271.mp4

**Comment by pklloveyou** [comment]
yes

**Comment by pklloveyou** [comment]
image
Here is the error. After the flight is completed and lands, it fails to clear the point successfully

**Comment by pklloveyou** [comment]
After deleting it, all functions work normally

**Comment by pklloveyou** [comment]
@Gultask

**Comment by Gultask** [comment]
Tested pkIloveyou's modification. The issue can be reproduced if you relog the moment right before you dismount, but it fixes the issue if you're already on the ground like in the video.

**Review by acidmanifesto — CHANGES_REQUESTED** [review]
got some chatter and a video of a potentional issue in the comments.

**Comment by UltraNix** [comment]
@pklloveyou Please. test it again.

**Comment by pklloveyou** [comment]
Tested
It has been used on my server for a long time, and there is no error

**Review by Nefertumm — APPROVED** [review]

**Review by acidmanifesto — APPROVED** [review]

**Comment by Nefertumm** [comment]
I'm posting here so it doesn't get lost. We had two crashes today on CC related to this commit, one was entering to rdf while in taxi and the other one the same but with bg. We kinda hotfixed that but dirty, making players to not be able to enter rdf/bg while on taxi.
Now there is another issue, which is the script is spawning a lot of waypoint creature, to the point it gets laggy after some time. I hope this could be fixed soon, since the only way to reduce the load right now is restarting the server.

File: src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp [hunk 2]
```diff
@@ -270,35 +270,71 @@ void WaypointMovementGenerator<Creature>::MovementInform(Creature* creature)
...
         TaxiPathNodeList const& nodes = sTaxiPathNodesByPath[path];
         if (!nodes.empty())
         {
+            TaxiPathNodeEntry const* start = nodes[0];
+            TaxiPathNodeEntry const* end = nodes[nodes.size() - 1];
+            bool passedPreviousSegmentProximityCheck = false;
             for (uint32 i = 0; i < nodes.size(); ++i)
-                i_path.push_back(nodes[i]);
+            {
+                sMapMgr->CreateMap(nodes[i]->mapid, player)->SummonCreature(1, { nodes[i]->x, nodes[i]->y, nodes[i]->z, 0.0f })->SetLevel(i ? i : 1);
```
Comment (Nefertumm) [inline src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp:318]: Does this creature have a purpose outside debugging?

Comment (pklloveyou) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp:318]: I tried it, and the code can work after deleting it

Comment (pklloveyou) [reply src/server/game/Movement/MovementGenerators/WaypointMovementGenerator.cpp:318]: and it
``` player->TeleportTo(node->map_id, node->x, node->y, node->z, player->GetOrientation()); ```

**Pull request merged**